fixed issue 802 by adding lastModified in error statement

nobellium1997 opened the pull request:
Super simple fix to a very simple issue listed. Thought this might be a good first try at contributing to open source.

Natim commented:
refs #802

Natim commented on src/collection.js:
```diff
@@ -989,7 +989,7 @@ export default class Collection {
     const emptyCollection = data.length === 0;
     if (!options.exclude && localSynced && serverChanged && emptyCollection) {
-      throw Error("Server has been flushed.");
+      throw Error("Server has been flushed." + since);
```
Thank you for your proposal. That's a good start.

However this is a bit too simple, no space between the dot and the value, we don't actually know what the since value is while reading the message. Is it the server timestamp or the client timestamp?

> I guess the lastModified from the client as well as the one the server responded with would be good.

Can you enhance your PR by adding commits to it and display both the server side timestamp and the client side one?

I would also recommend adding a test for it.

Natim replied:
The test you are looking to contribute to is here: https://github.com/Kinto/kinto.js/blob/571b0aeb2cc69f67d86523133d46b332cca5b3ae/test/integration_test.js#L2001

nobellium1997 replied:
Hi, so I fixed the issue with the client and the server timestamp, how would I go about matching the error to the error message in the integration_test file?

glasserc replied:
You can use a regular expression instead of a literal string to match the error message.

glasserc commented:
It also wouldn't be bad to expose the two values as properties of the error, although this is a bit more work!

nobellium1997 commented:
Hey guys, so I added regex to the integration_test file so that it matches on the phrase and the timestamps. I also added the new error message with both the server and client timestamps.

Natim commented:
Thanks the test looks great. I agree with @glasserc that it would be great to also add the ``serverTimestamp`` and ``localTimestamp`` properties to the ``Error`` object so that we can use them  if we catch the Error in some code.

Natim commented:
Thanks to the test we can see that ``serverChanged`` is not the server timestamp. We need to find it somewhere else apparently.

nobellium1997 commented:
Hey guys, so I just had a few questions. When you mean add the properties to the error object, where exactly should I be adding those properties? Looks like an error object only has a name, message, line number, and description so I'm not 100% sure where I would put the two properties. Also, I think the server timestamp is the variable named "unquoted" on line 983 in src/collection.js since it determines whether serverChanged is true or not.

nobellium1997 commented:
I can definitely work on generating a subclass, I'm on break so I have time :smile:. One more question, I can't get the integration tests to run when I run npm run test, for some reason it keeps saying something like "before all hook" followed by ncaught Error: spawn pserve ENOENT and I was wondering if there was an obvious fix. I tried looking online but no luck so far.

glasserc commented:
The integration tests try to start `kinto`, so make sure you have it in your `PATH`. I usually get that error message when I forgot to activate my virtualenv. (Of course, if setting up `kinto` locally is too much work, it's fine to just rely on travis.)

nobellium1997 commented:
Hey guys, I fixed the typo and made the client timestamp and server timestamp part of the error properties. I'll go ahead and start working on making a new error class now.

nobellium1997 commented:
Hey everyone, is this pr still in review? I think it should be good to go

Natim commented:
It looks good to me. r? @glasserc @leplatrem ?

nobellium1997 commented on src/collection.js:
```diff
@@ -989,7 +1002,17 @@ export default class Collection {
     if (!options.exclude && localSynced && serverChanged && emptyCollection) {
-      throw Error("Server has been flushed.");
+      // throw Error("Server has been flushed. Client Side Timestamp: " + localSynced + " Server Side Timestamp: "
+      // + unquoted);
```
Hey all, just removed it. Totally forgot I left that there

nobellium1997 commented:
Alrighty, I just removed the commented code so hopefully it should be good to go!

glasserc commented:
Thanks for your patience :)
